### PR TITLE
COR-396 Fix Single Select Field

### PR DESCRIPTION
**Pull request by ludydoo**
1. Moves the `FieldDefinition.Options` to `FieldTypeSingleSelect.Options`
2. Options are `[{"name":"option-name","id":"option-id"}]`
3. Records must specify options like `values:[{fieldId: field-id, value: option-id}]`. We do not put the `option name` in the field value, as this allows users to rename options without having to change the value of every record that references that option.
3. Added logic to create the underlying SQL structure to store select options
4. Added tests to make sure we can create forms with single select fields, as well as records for those forms
5. Added validation rules to the `FormDefinition` to validate a single select `FieldDefinition`
6. Added validation rules to the `Record` to validate incoming records that have values for a single select
7. Added flatten/hydrate methods in the form store to store/restore field definitions that contain single select fields
8. Simplified testing logic for the form store flatten/hydrate methods
9. Rename all instances of `core-js-api-client` to `core-api-client`
10. Fix pwa to use single selects with the new api structure

### CODE REVIEW

File: pkg/api/types/validation/form.go [hunk 1]
```diff
@@ -387,8 +394,46 @@ func ValidateFieldTypeQuantity(ftQuantity *types.FieldTypeQuantity, path *valida
...
+	if optionsCount == 0 {
+		// checking that we have at least one option
+		result = append(result, validation.Required(optionsPath, errSelectOptionsRequired))
+	} else if optionsCount > selectFieldMaxOptions {
+		// checking that we don't exceed the max option count
+		// we also don't process this further since this might be a crazy long list
+		return append(result, validation.TooMany(optionsPath, optionsCount, selectFieldMaxOptions))
+	}
+
+	seenNames := sets.NewString()
```
[the remark this answers is not in the document]
Comment (internetti) [reply pkg/api/types/validation/form.go:412]: shouldn't the names also be unique?

File: pkg/api/types/validation/record.go [hunk 2]
```diff
@@ -164,20 +164,21 @@ func ValidateRecordValue(path *validation.Path, value *string, field *types.Fiel
```
Comment (internetti) [inline pkg/api/types/validation/record.go:164]: for this kind of switches, what do you think about having a mapping object/config somewhere like
`
validators: {
  FieldKindText: ValidateRecordStringValue,
  FieldKindReference: ValidateRecordReferenceValue,
  ...
}
`
which then would shorten this to `validators[fieldKind](path, value, field)`

File: pkg/server/public/tests/form_test.go [hunk 3]
```diff
@@ -33,6 +33,11 @@ func (s *Suite) TestFormCreateGetList() {
 		return
 	}
 
+	selectOptions := []*types.SelectOption{
```
Comment (internetti) [inline pkg/server/public/tests/form_test.go:36]: is there a way in go to tag properties as required? if so, should the id of SelectOption not be required? since we're using typescript in the frontend, I think we would benefit of typesafety also in the backend

File: pkg/sqlmanager/build_single_select_field.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+// +-------+----------+
+//
+func singleSelectFieldActions(formInterface types.FormInterface, fieldDefinition *types.FieldDefinition) (sqlActions, error) {
+	result := sqlActions{}
+	result = append(result, buildSelectOptionsTable(formInterface, fieldDefinition)...)
+
+	sqlField := getStandardSQLColumnForField(fieldDefinition)
+	sqlField.DataType = schema.SQLDataType{
+		VarChar: &schema.SQLDataTypeVarChar{
+			Length: 36,
```
Comment (internetti) [inline pkg/sqlmanager/build_single_select_field.go:36]: could this 36 be moved to a variable, since it seems to be used in a lot of the field builders?

**Review by internetti — CHANGES_REQUESTED** [review]

File: .prow/pull-core-frontend-test.yaml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 postsubmits:
   - name: pull-core-frontend-test
-    run_if_changed: "^frontend"
+    run_if_changed: "frontend/.*"
```
[the remark this answers is not in the document]
Comment (ludydoo) [reply .prow/pull-core-frontend-test.yaml:3]: That should not happen for non-integration tests. When we do add cypress, that would be probably another prow job entirely

File: frontend/apps/native/src/components/screen_containers/AddRecordScreenContainer.tsx [hunk 6]
```diff
@@ -1,18 +1,18 @@
-import { FormDefinition } from "core-js-api-client";
-import { NetworkStateType } from "expo-network";
+import {FormDefinition} from "core-api-client";
```
[the remark this answers is not in the document]
Comment (ludydoo) [reply frontend/apps/native/src/components/screen_containers/AddRecordScreenContainer.tsx:1]: I do agree, I think Goland did some tricks on me for this one..

File: pkg/api/types/validation/form.go [hunk 7]
```diff
@@ -29,19 +30,25 @@ const (
...
+	errSelectOptionNameInvalid               = "Name of the select option is invalid"
 	fieldCodeMaxLength                       = 32
 	fieldCodeMinLength                       = 1
 	fieldNameMaxLength                       = 128
 	fieldNameMinLength                       = 2
 	formNameMaxLength                        = 128
 	formNameMinLength                        = 3
 	formMaxFieldCount                        = 60
+	selectFieldMaxNameLength                 = 128
+	selectFieldMaxOptions                    = 100
```
[the remark this answers is not in the document]
Comment (ludydoo) [reply pkg/api/types/validation/form.go:44]: I believed that countries would be a reference field..

File: pkg/api/types/validation/form.go [hunk 8]
```diff
@@ -29,19 +30,25 @@ const (
...
 	// todo: add maximum total number of fields (form fields + subform fields)
 )
 
 var (
-	fieldNameRegex = regexp.MustCompile("^[a-zA-Z0-9]+( [a-zA-Z0-9]+)*$")
-	fieldCodeRegex = regexp.MustCompile("^[a-zA-Z]+[a-zA-Z0-9]*$")
+	fieldNameRegex  = regexp.MustCompile("^[a-zA-Z0-9]+( [a-zA-Z0-9]+)*$")
+	fieldCodeRegex  = regexp.MustCompile("^[a-zA-Z]+[a-zA-Z0-9]*$")
+	optionNameRegex = regexp.MustCompile("^[a-zA-Z0-9]+( [a-zA-Z0-9]+)*$")
```
[the remark this answers is not in the document]
Comment (ludydoo) [reply pkg/api/types/validation/form.go:51]: Yeah, those are super restrictive regexes.. I think they are going to evolve as we do the seeders

File: pkg/api/types/validation/form.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ludydoo) [reply pkg/api/types/validation/form.go:412]: This is for asserting the uniqueness of names, if the set already contains a name, that means it is a duplicate one, there's a validation error thrown for duplicate option names

File: pkg/api/types/validation/record.go (hunk 2 above)
Comment (ludydoo) [reply pkg/api/types/validation/record.go:164]: Absolutely. we should probably have an interface for field types. Because there are some stuff that fields always do
- Validate the record value
- Validate the form definition
- Convert to SQL representation
- Convert from SQL representation
- Add SQL fields/tables for storing the field
All that stuff would probably need to be under an interface such as
```
type FieldStrategy interface{
  ValidateRecordValue(value ...) validation.ErrorList
  ValidateFieldDefinition(fieldDefinition *types.FieldDefinition) validation.ErrorList
  ConvertToSQL(value ..) (interface{}, error)
  ConvertFromSQL(value ..) (FieldValue, error)
  PrepareSQLInfra(?) error
}

func GetStrategyForField(fieldDefinition *types.FieldDefinition) (FieldStrategy, error) {...}
```

File: pkg/server/public/tests/form_test.go (hunk 3 above)
Comment (ludydoo) [reply pkg/server/public/tests/form_test.go:36]: In go, not really. It's possible to create any kind of struct without passing any fields. For types that are instantiated in the code, then we use a constructor with required arguments
```
NewFieldOption(id string, value..)
```
In this case, the type is deserialized, so there is no built-in way of doing this at compile time. The validator package would take care of that..

File: pkg/sqlmanager/build_single_select_field.go (hunk 4 above)
Comment (ludydoo) [reply pkg/sqlmanager/build_single_select_field.go:36]: Yes, will do that

File: .prow/pull-core-frontend-test.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (internetti) [reply .prow/pull-core-frontend-test.yaml:3]: but when you changed the subform structure in the api, it broke the pwa tests

[the remark this answers is not in the document]
Comment (ludydoo) [reply .prow/pull-core-frontend-test.yaml:3]: The frontend tests don't make any api calls. The changes that probably broke the tests were the one in `/frontend/packages/client` I think

File: pkg/api/types/validation/form.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (internetti) [reply pkg/api/types/validation/form.go:51]: so you mean it's not worth it doing that yet?

File: pkg/api/types/validation/record.go (hunk 2 above)
Comment (internetti) [reply pkg/api/types/validation/record.go:164]: I'll create a ticket for it

File: frontend/apps/pwa/src/features/former/former.slice.ts [hunk 9]
```diff
@@ -153,6 +153,8 @@ const mapFields = (state: FormerState, fields: FormField[]): FieldDefinition[] =
 
     for (let field of fields) {
 
+        console.log("FIELD", field)
+
```
[the remark this answers is not in the document]
Comment (internetti) [reply frontend/apps/pwa/src/features/former/former.slice.ts]: please remove this

File: pkg/api/types/validation/form.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ludydoo) [reply pkg/api/types/validation/form.go:51]: I don't think it's worth it, it's probablyu going to happen with the seeder once we rebase

**Review by internetti — APPROVED** [review]
